odl/solvers/smooth/nonlinear_cg.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2014-2017 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Simple iterative type optimization schemes."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+
+from odl.solvers.util import ConstantLineSearch, BacktrackingLineSearch
+
+
+__all__ = ('conjugate_gradient_nonlinear',)
+
+
+def conjugate_gradient_nonlinear(f, x, line_search=1.0, maxiter=1000, nreset=0,
+                                 tol=1e-16, beta_method='FR',
+                                 callback=None):
+    """Conjugate gradient for nonlinear problems.
+
+    Notes
+    -----
+    This is a general and optimized implementation of the nonlinear conjguate
+    gradient method for solving a general unconstrained optimization problem
+
+        :math:`\min f(x)`
+
+    for a differentiable function
+    :math:`f: \mathcal{X}\\to \mathbb{R}` on a Hilbert space
+    :math:`\mathcal{X}`. It does so by finding a zero of the gradient
+
+        :math:`\\nabla f: \mathcal{X} \\to \mathcal{X}`.
+
+    The method is described in a
+    `Wikipedia article
+    <https://en.wikipedia.org/wiki/Nonlinear_conjugate_gradient_method>`_.
+
+    Parameters
+    ----------
+    f : `Functional`
+        Functional with ``f.gradient``.
+    x : ``op.domain`` element
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, uses it as a
+        fixed step length.
+    maxiter : int
```

Comment:
I guess because the algorithm can terminate early and is not guaranteed to run this number of iterations. Should be cross-checked anyway and changed in the other methods if it applies there, too. I like the distinction.